synapse/storage/databases/main/stream.py

```diff
@@ -208,6 +210,55 @@ def _make_generic_sql_bound(
     )
 
 
+def _filter_results(
+    lower_token: Optional[RoomStreamToken],
+    upper_token: Optional[RoomStreamToken],
+    instance_name: str,
+    topological_ordering: int,
+    stream_ordering: int,
+) -> bool:
+    """Returns True if the event persisted by the given instance at the given
+    topological/stream_ordering falls between the two tokens (taking a None
+    token to mean unbounded).
+
+    Used to filter results from fetching events in the DB against the given
+    tokens. This is necessary to handle the case where the tokens include
+    position maps, which we handle by fetching more than necessary from the DB
+    and then filtering (rather than attempting to construct a complicated SQL
+    query).
+    """
+
+    event_historical_tuple = (
+        topological_ordering,
+        stream_ordering,
+    )
+
+    if lower_token:
+        if lower_token.topological:
+            # If these are historical tokens we compare the `(topological, stream)`
+            # tuples.
+            if event_historical_tuple <= lower_token.as_historical_tuple():
+                return False
+
+        else:
+            # If these are live tokens we compare the stream ordering against the
+            # writers stream position.
+            if stream_ordering <= lower_token.get_stream_pos_for_instance(
+                instance_name
+            ):
+                return False
```

Comment:
I wonder if we should lift these comparisons into `PersistedEventPosition` (and give `PersistedEventPosition` the event's topological ordering), then we could just do `lower_token < event_pos < upper_token`?

Reply:
One problem with overloading `<` in that way is that, given the different ways we have of comparing token orderings (ref the existence of `PersistedEventPosition.persisted_after`), `<` fails to make it clear which mechanism we are using.

So yes, I think it might be nice to move the comparison into `PersistedEventPosition`, but it might be better to do it with explicit methods rather than `<`.

(also, note that `lower_token < event_pos` would call `RoomStreamToken.__lt__`, so you'd have to put logic in `RoomStreamToken` too)

Reply:
(I believe `persisted_after` has the same semantics)